docs/standard/serialization/binaryformatter-migration-guide/reading-nrbf-payloads.md

```diff
@@ -0,0 +1,236 @@
+
+All `[Serializable]` types from [Quartz.NET](https://github.com/search?q=repo%3Aquartznet%2Fquartznet+%5BSerializable%5D+language%3AC%23&type=code&l=C%23) that can be persisted by the library itself are `sealed`, so there are no custom types that the users could create and the payload can contain only known types. They also provide public constructors, so it is possible to re-create these types based on the information read from payload.
+
+`SettingsPropertyValue` type from `System.Configuration.ConfigurationManager` library exposes an `object PropertyValue` that may internally use `BinaryFormatter` to serialize and deserialize any object that was stored in the configuration file. It could be used to store an integer, a custom type, a dictionary or literally anything.  Because of that, **it is impossible to migrate this library without introducing breaking changes to the API**.
+
+### Identifying BinaryFormatter payload
+
+`NrbfDecoder` provides two `StartsWithPayloadHeader` methods that allow to check whether given stream or buffer starts with NRBF header.
+
+Using these methods is recommended for the period of migrating payloads persisted with `BinaryFormatter` to a [different serializer](./choosing-a-serializer.md):
+
+- Check if the payload read from storage is an [NRBF](/openspecs/windows_protocols/ms-nrbf/75b9fe09-be15-475f-85b8-ae7b7558cfe5) payload.
+- If so, read it with `NrbfDecoder`, serialize it back with a new serializer and overwrite the data in the storage.
+- If not, use the new serializer to deserialize the data.
+
+```cs
+internal static T LoadFromFile<T>(string path)
+{
+    bool update = false;
+    T value;
+
+    using (FileStream stream = File.OpenRead(path))
+    {
+        if (NrbfDecoder.StartsWithPayloadHeader(stream))
+        {
+            value = LoadLegacyValue<T>(stream);
+            update = true;
+        }
+        else
+        {
+            value = LoadNewValue<T>(stream);
+        }
+    }
+
+    if (update)
+    {
+        File.WriteAllBytes(path, NewSerializer(value));
+    }
+
+    return value;
+}
+```
+
+### Safely reading NRBF payloads
+
+The NRBF payload consists of serialization records that represent the serialized objects and their metadata. To read the whole payload and get the root object, the user needs to call the `Decode` method.
+
+The `Decode` method returns a `SerializationRecord` instance. `SerializationRecord` is an abstract class that represents the serialization record and provides three self-describing properties: `Id`, `RecordType` and `TypeName`. It exposes one method: `public bool TypeNameMatches(Type type)` which compares the type name read from the payload (and exposed via `TypeName` property) against the specified type. This method ignores assembly names, so the users don't need to worry about type forwarding and assembly versioning. It also does not consider member names or their types (because getting this information would require type loading).
```

Comment:
```suggestion
The `Decode` method returns a `SerializationRecord` instance. `SerializationRecord` is an abstract class that represents the serialization record and provides three self-describing properties: `Id`, `RecordType`, and `TypeName`. It exposes one method, `public bool TypeNameMatches(Type type)`, which compares the type name read from the payload (and exposed via `TypeName` property) against the specified type. This method ignores assembly names, so users don't need to worry about type forwarding and assembly versioning. It also does not consider member names or their types (because getting this information would require type loading).
```